Add basic support for s390x

uweigand opened the pull request:
This adds basic support for Linux on s390x, using the fallback implementation of all algorithms and no assembler routines.   This addresses https://github.com/briansmith/ring/issues/986 and makes "cargo test" fully pass on s390x.

There were two main changes I had to make to get this working:

- Provide a bn_mul_mont fallback implementation.  Note that while there is a fallback implementation on the Rust side for limbs_mont_mul, some elliptic curve code directly calls bn_mul_mont on the C side, where there is currently no fallback.  However, given that we have a bn_from_montgomery_in_place fallback, it is straightforward to add bn_mul_mont (along the lines of the limbs_mont_mul fallback).
- Support big-endian platforms.   This consists of adding a number of endian access primitives in internal.h, and using them where required, currently in aes_nohw.c, poly1305.c, and the p256_scalar_bytes_from_limbs routine (for some reason, the p384 version of that routine is already endian-agnostic).  I have chosen to use a plain C implementation of the primitives rather than bswap intrinsics, because that is easier to maintain, doesn't require platform #ifdef's, and still compiles to good code with any recent compiler.

@briansmith, please let me know if this approach makes sense to you or if this should be done differently.

Once this basic (unoptimized) support is in, we can then build on it by adding assembler optimizations (they're already present in OpenSSL, but would need to be copied over and adapted).

uweigand commented:
Ping?  Any comments on this approach?

briansmith commented on crypto/internal.h:
```diff
@@ -254,19 +254,77 @@ static inline crypto_word constant_time_select_w(crypto_word mask,
-#if defined(__GNUC__) && __GNUC__ >= 2
-static inline uint32_t CRYPTO_bswap4(uint32_t x) {
-  return __builtin_bswap32(x);
+static inline uint32_t CRYPTO_read_le32(const uint8_t *p) {
```
I will do a merge from the BoringSSL sources to include these functions.

uweigand replied:
Note that the BoringSSL versions (CRYPTO_load_u32_le etc.) only work correctly on little-endian hosts.  (I guess this goes back to the BoringSSL team's decision to not support big-endian hosts at all.)

The versions I've provided in this PR work on any host, little-endian or big-endian.

briansmith replied:
Yes, I know if I merge the BoringSSL versions then they'll need to be modified to support big endian. I also am very fond of the endian-neutral implementation style you chose. However, I remember that one compiler I tested (I think MSVC) didn't recognize those idioms as of a couple years ago. Also, it's easier to make use of the BoringSSL team's performance/benchmarking work and the AWS/Galois team's verification work if we keep the code the same for little-endian platforms. That's why I'm thinking to merge in the BoringSSL teams's functions and then we can modify them to support big-endian.

uweigand replied:
Ok, I see.  That should work as well, thanks.

briansmith commented on crypto/fipsmodule/ec/p256_shared.h:
```diff
@@ -50,7 +50,14 @@ typedef unsigned char P256_SCALAR_BYTES[33];
 
 static inline void p256_scalar_bytes_from_limbs(
     P256_SCALAR_BYTES bytes_out, const BN_ULONG limbs[P256_LIMBS]) {
-  OPENSSL_memcpy(bytes_out, limbs, 32);
```
I need to go through all the remaining uses of OPENSSL_memcpy and make sure they're OK.

briansmith commented on crypto/fipsmodule/bn/montgomery.c:
```diff
@@ -156,3 +156,17 @@ int bn_from_montgomery_in_place(BN_ULONG r[], size_t num_r, BN_ULONG a[],
+
+#if !defined(OPENSSL_X86) && !defined(OPENSSL_X86_64) && \
+    !defined(OPENSSL_ARM) && !defined(OPENSSL_AARCH64)
+void bn_mul_mont(BN_ULONG *rp, const BN_ULONG *ap, const BN_ULONG *bp,
```
This would need to have the prefixing stuff done to it in build.rs.

uweigand replied:
As far as I can see, bn_mul_mont is already in SYMBOLS_TO_PREFIX in build.rs ?

briansmith replied:
Thanks for pointing that out.

briansmith commented on crypto/fipsmodule/bn/montgomery.c:
```diff
@@ -156,3 +156,17 @@ int bn_from_montgomery_in_place(BN_ULONG r[], size_t num_r, BN_ULONG a[],
+    !defined(OPENSSL_ARM) && !defined(OPENSSL_AARCH64)
+void bn_mul_mont(BN_ULONG *rp, const BN_ULONG *ap, const BN_ULONG *bp,
+                 const BN_ULONG *np, const BN_ULONG *n0, size_t num) {
+  Limb tmp[2 * num];
```
I agree this makes sense. This would be the first place we've used variable-length arrays in *ring* though, which makes me hesitate. I will circle back to this.

uweigand replied:
I guess we could use a global maximum limit along the lines of MODULUS_MAX_LIMBS in src/arithmetic/bigint.rs.   I'd be happy to implement whatever change you prefer -- just let me know.

vimalk78 commented:
hi @uweigand @briansmith, i am also interested in getting `ring` supported for s390x. Let me know if i can contribute in some way.
From this PR, I was able to build `ring` using `cross build --release --target s390x-unknown-linux-gnu`
btw, how do we get the above tests to run on the PR?

uweigand commented:
Hi @briansmith, let me try to make another attempt to get this going.

I've just updated the patch to use a different style of endian conversion, as you suggested.   This keeps the CRYPTO_bswap4 implementation from boringssl and adds the corresponding CRYPTO_bswap8 implementation from there as well.   The other endian-specific accessors are now all defined in term of CRYPTO_bswap4 / CRYPTO_bswap8 plus OPENSSL_memcpy (to handle any alignment issues if necessary).

Does this address your concerns?

If there's anything else I can do to help get this reviewed and merged, please let me know.

uweigand commented:
Fixed merge conflict against current mainline.   Any comments?

Xynnn007 commented:
Hi @briansmith , can this pr be merged now?

uweigand commented:
Rebased and fixed merge conflicts against current mainline.

uweigand commented:
Added support for cross-building and testing as requested in https://github.com/briansmith/ring/issues/1555.
Split into three logically distinct commits.

uweigand commented:
PR updated after https://github.com/briansmith/ring/pull/1558 was merged.

ecnelises commented:
This works well and can be the foundation of more big endian targets (like AIX). Any blockers?

tnakaike commented:
I am expecting that this PR is merged because many components of [confidential containers](https://github.com/confidential-containers), which support s390x, are written in rust. For example, [Key Broker Service](https://github.com/confidential-containers/kbs) uses rustls, which depends on ring, to enable https.

angelnu commented:
As Ring is used by [PRISMA](https://www.prisma.io) we would depend on this patch to be able to use Prima on s390x. As we (Metaco and Ripple) offer multi-arch builds for our SW this would help us a lot. We are looking at moving more components to Rust so this crypto library being available in s390x would be very appreciated.

uweigand commented:
> @uweigand Could you (or whoever is working on this at IBM/RedHat) email me at [brian@briansmith.org](mailto:brian@briansmith.org) so we could coordinate how we could get this finished without it being blocked on me, and how s390x (and PPC?) support can be best maintained going forward. I have some ideas and I want to hear your ideas.

I've now reached out to you via email.

ecnelises commented:
Hi @uweigand could you please CC me? AIX stuff may be related, thanks.

briansmith commented on crypto/fipsmodule/aes/aes_nohw.c:
```diff
@@ -338,15 +338,15 @@ static inline void aes_nohw_compact_block(aes_word_t out[AES_NOHW_BLOCK_WORDS],
 #if defined(OPENSSL_SSE2)
   // No conversions needed.
 #elif defined(OPENSSL_64_BIT)
-  uint64_t a0 = aes_nohw_compact_word(out[0]);
```
I noticed that you have added `CRYPTO_bswap8_le` to every invocation of `aes_nohw_compact_word`. Instead, you could move the `CRYPTO_bswap8_le` into `aes_nohw_compact_word` itself.

Further, if you look at `aes_nohw_compact_word`, you can see that `aes_nohw_compact_word` looks like this:

```
  // Numbering the 64/2 = 16 4-bit chunks, least to most significant, we swap
  // quartets of those chunks:
  //   0 1 2 3 | 4 5 6 7 | 8  9 10 11 | 12 13 14 15 =>
  //   0 2 1 3 | 4 6 5 7 | 8 10  9 11 | 12 14 13 15
  a = aes_nohw_delta_swap(a, UINT64_C(0x00f000f000f000f0), 4);
  // Swap quartets of 8-bit chunks (still numbering by 4-bit chunks):
  //   0 2 1 3 | 4 6 5 7 | 8 10  9 11 | 12 14 13 15 =>
  //   0 2 4 6 | 1 3 5 7 | 8 10 12 14 |  9 11 13 15
  a = aes_nohw_delta_swap(a, UINT64_C(0x0000ff000000ff00), 8);
  // Swap quartets of 16-bit chunks (still numbering by 4-bit chunks):
  //   0 2 4 6 | 1  3  5  7 | 8 10 12 14 | 9 11 13 15 =>
  //   0 2 4 6 | 8 10 12 14 | 1  3  5  7 | 9 11 13 15
  a = aes_nohw_delta_swap(a, UINT64_C(0x00000000ffff0000), 16);
```
So if you really wanted to, I think you might be able to change the inputs to the aes_nohw_delta_swap calls to optimize out the `CRYPTO_bswap8_le` call completely. If you want to save this change for the future, we can defer this change.

But at a minimum, let's move the byte swapping to `aes_nohw_compact_word` to make the diff from BoringSSL smaller.

uweigand replied:
Given your later comments, I've now changed this to use CRYPTO_load/store instead.  I think this makes the rest of the comments here obsolete - please let me know if you'd like me to do anything different here.

briansmith replied:
Hi, I would rather we keep this code exactly as it is in BoringSSL as much as possible and instead add the `CRYPTO_bswap` call to the top of `aes_nohw_compact_word` along with a`"// TODO: merge byte swapping into delta swapping?` comment, and the same for `uncompact_word`.  This way we ensure we're not impeding the performance at all on the already-supported targets. I guess this is the only two places where we'll need to use `CRYPTO_bswap` outside internal.h, and hopefully it is temporary (NBD if not).

briansmith commented on crypto/fipsmodule/ec/p256_shared.h:
```diff
@@ -51,7 +51,20 @@ typedef unsigned char P256_SCALAR_BYTES[33];
+    OPENSSL_memcpy(bytes_out + i * 4, &limb, 4);
+#endif
+  }
+#else
```
In the recent merge from BoringSSL I brought in the following functions:

```
static inline crypto_word CRYPTO_load_word_le(const void *in) {
  crypto_word v;
  OPENSSL_memcpy(&v, in, sizeof(v));
  return v;
}

static inline void CRYPTO_store_word_le(void *out, crypto_word v) {
  OPENSSL_memcpy(out, &v, sizeof(v));
}
```
Please use these functions to eliminate the conditional compilation logic here.

uweigand replied:
On re-testing I noticed that all this code is never even used on s390x these days - the whole file is only compiled in on x86_64 and aarch64 targets.  Therefore I've just reverted this change now.

briansmith commented on crypto/internal.h:
```diff
@@ -264,6 +264,10 @@ static inline crypto_word constant_time_select_w(crypto_word mask,
+
+static inline uint64_t CRYPTO_bswap8(uint64_t x) {
+  return __builtin_bswap64(x);
+}
```
This was brought in by the recent BoringSSL merge.

uweigand replied:
Thanks!

briansmith commented on crypto/internal.h:
```diff
@@ -272,8 +276,44 @@ static inline uint32_t CRYPTO_bswap4(uint32_t x) {
+
+static inline uint64_t CRYPTO_bswap8(uint64_t x) {
+  return CRYPTO_bswap4(x >> 32) | (((uint64_t)CRYPTO_bswap4(x)) << 32);
+}
```
I think these are all covered bythe recent BoringSSL merge too.

Note that `CRYPTO_bswap8` and `CRYPTO_bswap4` should probably only be used for implementing the functions mentioned below; they shouldn't be used outside of internal.h, AFAICT.

uweigand replied:
Agreed.

briansmith commented on crypto/internal.h:
```diff
@@ -272,8 +276,44 @@ static inline uint32_t CRYPTO_bswap4(uint32_t x) {
+#define CRYPTO_bswap8_le(x) (x)
+#define CRYPTO_bswap4_be(x) CRYPTO_bswap4((x))
+#define CRYPTO_bswap8_be(x) CRYPTO_bswap8((x))
+#endif
```
In general we don't use `#define` for functions; instead we use `static inline` functions. To help you, I brought in the BoringSSL definitions of the following functions in the recent BoringSSL merge:
```
static inline uint32_t CRYPTO_load_u32_le(const void *in);
static inline void CRYPTO_store_u32_le(void *out, uint32_t v);
static inline uint32_t CRYPTO_load_u32_be(const void *in);
static inline void CRYPTO_store_u32_be(void *out, uint32_t v);
static inline uint64_t CRYPTO_load_u64_le(const void *in);
static inline void CRYPTO_store_u64_le(void *out, uint64_t v);
static inline uint64_t CRYPTO_load_u64_be(const void *ptr);
static inline void CRYPTO_store_u64_be(void *out, uint64_t v);
static inline crypto_word_t CRYPTO_load_word_le(const void *in);
static inline void CRYPTO_store_word_le(void *out, crypto_word_t v);
static inline crypto_word_t CRYPTO_load_word_be(const void *in);
```
Please use these functions throughout the PR as they are more consistent with the Rust equivalents.

Basically I think your PR will mostly reduce to implementing the big-endian versions of these functions.

Note that some of these functions may be dead code in *ring* now and you may just remove the dead ones instead of porting them to big-endian, if you want.

uweigand replied:
I've removed the load_word / store_word versions as they are unneded.  Everything else I've added big-endian versions for.

briansmith commented on crypto/poly1305/poly1305.c:
```diff
@@ -29,14 +29,14 @@
 static void U32TO8_LE(uint8_t *m, uint32_t v) {
+  v = CRYPTO_bswap4_le(v);
   OPENSSL_memcpy(m, &v, sizeof(v));
 }
```
The recent BoringSSL merge fixed this.

uweigand replied:
Fixed.

briansmith commented on mk/cargo.sh:
```diff
@@ -108,6 +109,13 @@ case $target in
+    export CFLAGS_s390x_unknown_linux_gnu="--sysroot=/usr/s390x-linux-gnu"
+    export CARGO_TARGET_S390X_UNKNOWN_LINUX_GNU_LINKER=s390x-linux-gnu-gcc
+    export CARGO_TARGET_S390X_UNKNOWN_LINUX_GNU_RUNNER="$qemu_s390x"
+    ;;
```
Does clang work? Do you have a preferred toolchain, GCC vs clang? Generally I prefer clang because it is much lighter weight. If your team has an official preference for GCC then we can keep using GCC.

briansmith replied:
I mean, is there a way to avoid using `s390x-linux-gnu-gcc` for linking and instead use rust-lld or clang as the linker?

uweigand replied:
lld doesn't support the s390x target at this point.  It should be possible to use the clang driver to invoke the (system) linker, but that'll just do the same as the gcc driver.  Do you want me to make that change anyway?

briansmith replied:
> lld doesn't support the s390x target at this point. It should be possible to use the clang driver to invoke the (system) linker, but that'll just do the same as the gcc driver. Do you want me to make that change anyway?

No, it's OK this way. Thanks!

briansmith reviewed (changes requested):
I am sorry for the long delay here. I have now done enough of the BoringSSL merge where we can move forward with this. I am sorry so many changes are necessary but the suggested changes will reduce work in the long run.

briansmith commented:
I am planning to do a release very soon, probably later this week. If you can update your PR then I will include it in the release. Thanks for your help here.

briansmith commented:
> I am planning to do a release very soon, probably later this week. If you can update your PR then I will include it in the release. Thanks for your help here.

Otherwise, if we do the 0.17 release without these changes, we can do a 0.17.1 with them this gets merged.

uweigand commented on crypto/fipsmodule/aes/aes_nohw.c:
```diff
@@ -392,6 +394,10 @@ static inline void aes_nohw_uncompact_block(
+  b0 = CRYPTO_bswap4_le(b0);
+  b1 = CRYPTO_bswap4_le(b1);
+  b2 = CRYPTO_bswap4_le(b2);
+  b3 = CRYPTO_bswap4_le(b3);
```
See above.

uweigand commented on crypto/fipsmodule/aes/aes_nohw.c:
```diff
@@ -920,11 +926,11 @@ void aes_nohw_ctr32_encrypt_blocks(const uint8_t *in, uint8_t *out,
     // Update counters.
     for (uint32_t i = 0; i < AES_NOHW_BATCH_SIZE; i++) {
-      ivs.u32[4 * i + 3] = CRYPTO_bswap4(ctr + i);
+      ivs.u32[4 * i + 3] = CRYPTO_bswap4_be(ctr + i);
```
Fixed.

uweigand commented:
I think I've addressed all comments above.  Let me know if you would like me to change anything else.

briansmith commented on crypto/fipsmodule/aes/aes_nohw.c:
```diff
@@ -392,10 +391,10 @@ static inline void aes_nohw_uncompact_block(
+  CRYPTO_store_u32_le(out, b0);
+  CRYPTO_store_u32_le(out + 4, b1);
+  CRYPTO_store_u32_le(out + 8, b1);
+  CRYPTO_store_u32_le(out + 23, b1);
```
Ditto here.

uweigand replied:
OK, changed to use `CRYPTO_bswap` now.  Note I don't quite see how this could be folded completely into the existing delta swapping calls, it seems to me we'd need at least one extra call.  In that case, just doing the byteswap separately is likely to be more efficient (as big-endian targets are likely to have an optimized byteswap instruction).

briansmith commented on src/lib.rs:
```diff
@@ -115,5 +115,3 @@ mod sealed {
 }
 
-// TODO: https://github.com/briansmith/ring/issues/1555.
-const _LITTLE_ENDIAN_ONLY: () = assert!(cfg!(target_endian = "little"));
```
Please leave this in, but add:
```
// XXX: 64-bit big endian is tested; 32-bit is not. 
// TODO: Add 32-bit big endian test coverage to CI.
const _ENDIAN_TESTING: () = assert!(cfg!(any(target_endian = "little", target_pointer_width = "64")));
```

briansmith replied:
Also, `cargo fmt`.

uweigand replied:
Done.

briansmith reviewed (changes requested):
Thank you! The thing I notice now is that the coverage reporting complains that the generic code isn't being picked up. This is because the coverage measurement is actually done in separate jobs in a separate test matrix.

In https://github.com/rust-lang/rust/pull/104304 you added profiler support for this target. So I think we can add this target to the `coverage` CI job's matrix. We already support `aarch64-unknown-linux-gnu` in CI through QEMU so I believe it should be easy to add this.

briansmith commented:
Note that you did all the work to enable s390x in CI but you didn't add s390x to the `test`, `test-features`, and `coverage` build matrices. We should do that so we have the right test coverage.

briansmith commented:
I will fix the --no-default-features build failure.

briansmith commented:
Please rebase on top of main and resubmit.

briansmith commented:
Why did the earlier version of this have changes to `p256_shared.h` but the newest one doesn't?

uweigand commented:
> Please rebase on top of main and resubmit.

Done.

> Why did the earlier version of this have changes to `p256_shared.h` but the newest one doesn't?

Huh, I thought I added a comment but it seems to have disappeared.   I've removed that change as re-testing showed it is no longer necessary in the current code base.

The change was to fix an endian issue in `p256_scalar_bytes_from_limbs`, but that function is now only called from the `crypto/fipsmodule/ec/p256-nistz.c` file, which in turn is only even built at all on the (little-endian) AARCH64 and X86_64 targets.  So there is no need for this file to support big-endian platforms, and any code added there would be untested anyway.

briansmith commented:
Good news:
The refactoring to fix the --no-default-features build was merged into main.

Less good news:
There's a merge conflict in crypto/internal.h from the latest BoringSSL merge. Also, PR #1663 moved the stuff in base.h to target.h. I will merge PR #1663 ASAP so you can rebase this on top of main for the release. I am hoping to finish the BoringSSL merge today so there shouldn't be any new conflicts.

uweigand commented:
@briansmith: Rebased against current mainline to fix both merge conflicts.

uweigand commented:
@briansmith - this is unfortunate.  The recently added `test_constant_time` test now fails on s390x.  However, the test passes when run on a native machine, and it turns out this is actually a bug in qemu - and one that was already fixed here: https://lists.gnu.org/archive/html/qemu-devel/2023-05/msg06965.html

For some reason, clang chooses to use this particular z13 instruction (LOCFHR - which is generally quite rare) when implementing the `bssl_constant_time_test_main` routine.  

I'm wondering what the best way to proceed here would be.   I can imagine a number of options:
- Just disable this test on s390x (either generally, or only when running on qemu)
- Use a qemu that has the above fix - but that would mean building qemu yourself, or switching to Ubuntu 23.04 (if github actions already support that?), as older distro versions aren't recent enough
- Tell clang to not emit this instruction, e.g. by selecting an older target architecture (adding `-march=zEC12` to `CFLAGS_s390x_unknown_linux_gnu` works for me)

Any preferences?

briansmith commented:
> Just disable this test on s390x (either generally, or only when running on qemu)

Let's pick this choice. Note that this test isn't new. It is one of the oldest ones, which is why it is written in C!

uweigand commented:
> > Just disable this test on s390x (either generally, or only when running on qemu)
> 
> Let's pick this choice. Note that this test isn't new. It is one of the oldest ones, which is why it is written in C!

Huh, interesting.  Not sure what exactly triggered this to show up just now - the test seems to have passed before the recent boringssl merges.  Anyway, with things like that, some random changes in surrounding code may trigger different instruction selection choices.

I've updated the patch now to exclude this one test on s390x.

briansmith commented:
Thanks. Maybe IBM can send me one of these 390x boxes? Then my daughter could do her homework on it when I'm not using it for testing.

uweigand commented:
> Thanks. Maybe IBM can send me one of these 390x boxes? Then my daughter could do her homework on it when I'm not using it for testing. :)

Well, we're not usually sending out boxes, but if you're actually interested, you could in fact get free access to s390x machines for the purpose of open-source development and testing :-)

See here: https://www.ibm.com/community/z/open-source/virtual-machines-request/

uweigand commented:
Rebased once again after the riscv64 merge.

briansmith commented on src/constant_time.rs:
```diff
@@ -40,6 +40,7 @@ mod tests {
     use crate::{bssl, error};
 
     #[test]
+    #[cfg(not(target_arch = "s390x"))] // Triggers a qemu bug before 8.0.3
```
OK, this is not what I was expecting. I thought we'd have an `#ifdef` around a line or two within `constant_time_test.cc`.

Maybe instead we should change ci.yml so that we have something like:
```
          - target: s390x-unknown-linux-gnu
            host_os: ubuntu-22.04
            # XXX: `constant_time_test` fails when run under QEMU 8.0.2 and earlier
            # (https://link that you shared with me in your earlier comment).
            cargo_options:  -- --skip constant_time_test
```

This way people would not ever run into the issue on bare metal testing and I could also play with it locally. WDYT?

uweigand replied:
I see, this is indeed nicer.  I've changed the PR accordingly.

briansmith commented on .github/workflows/ci.yml:
```diff
@@ -266,6 +267,9 @@ jobs:
             host_os: ubuntu-22.04
 
+          - target: s390x-unknown-linux-gnu
+            host_os: ubuntu-22.04
```
In the analogous PPC64LE PR I just merged, the author noted that they kept the alphabetical order intact. Could you please do the same throughout this PR? It's an arbitrary rule but it avoids people needing to wonder where to put things.

uweigand replied:
Used alphabetic order everywhere now.

briansmith reviewed (changes requested):
Your PR is the next one to merge after the requested changes are made. Sorry about the many conflicts and delays.

briansmith commented:
It seems like codecov doesn't understand s390x profiler output? Too bad, but we'll deal.

uweigand commented:
> It seems like codecov doesn't understand s390x profiler output? Too bad, but we'll deal.

Why do you say it doesn't understand the output?  I saw changes reported for most of the new lines added (except for those in unused inline functions), and also for the now-skipped test_constant_time test ...

uweigand commented:
Ah, the `--skip test_constant_time` unfortunately seems to break the full run:
```
     Running `qemu-s390x -L /usr/s390x-linux-gnu /home/uweigand/ring/target/s390x-unknown-linux-gnu/release/deps/aead-0c8df580230e6a2e --skip test_constant_time`
error: Found argument '--skip' which wasn't expected, or isn't valid in this context
```
It is passed to both tests and benchmarks, and benchmarks apparently do not expect the `--skip` argument.

I've now changed the patch to use `-march=zEC12` instead.   That seems the straightforward workaround (directly addresses the root cause that QEMU fails to correctly emulate a z13 instruction), only needs a change in one place, and also is used only when doing the QEMU-based cross testing.

What do you think?   If you have any other suggestion, please let me know.

briansmith commented:
Thanks!

uweigand commented:
Thanks for your support in getting this merged!
